### DISCUSSION TITLE
Is next.js a good choice when building a front end for an existing backend api

### DISCUSSION THREAD

**Discussion by sam2000** [vercel/next.js#66291] [Help] [answered]
### Summary

I understand that next.js is a full stack solution.  I have already built an api (it is an AWS backend using Cognito, API Gateway, dynamoDB etc.). So all I need is a client. So I guess I could use plain react or  create react app. Wondering if next.js is a good fit for such a client-only solution. I would like to find out
1. Is it uncommon or unusual to use next.js in such a client-only case?
2. How does it compare to using 'create react app' or 'plain react' in such a use case?

I am mainly interested in speed/ease of development on the client side and I would like to choose something that is easy to use to build the client. So wanted to find out if next.js is a good choice in that respect.

**Comment by icyJoseph** [comment, marked as answer]
Next.js, as a "full stack solution" 🤔  I think many people would argue with you about that. 

Most commonly accepted is that, Next.js is a backend for your frontend. Or rather, that the division of responsibilities is somewhat altered. 

Next.js can take care of your frontend, while also providing a server side runtime/space, which, can in turn communicate with your data sources. It'd rather think of it, as an umbrella runtime, that can accommodate part of the full web stack, but not all of it.

To keep the discussion within bounds of your question, you have a trade-off to consider.

If you go the vite route (avoid CRA, or just plain setting up everything yourself), parcel, etc, you'd need to consider:

- Do you want to set isomorphic rendering yourself, or just ship an empty HTML shell to do full CSR 
- You'd have to consider that CSR has to do a few steps first, before it can go and get the data from your already existing DB
- Your already existing solution, to some degree will have to have public endpoints
- Routing has to be done properly, serving a 200, but then having a client side router render a "Not found" page, is not wrong, but not great either, redirects this way are also not a great idea
- if you need to join data with other services, you'd need to do that client side, which might require, loading the client with API keys and such, that are often sensitive and are often subject of debate
- ...etc

If you do go the Next.js route, you need to consider:

- Increased latency in the total round trip of a request from client to next.js to your already existing server
   - However, if you are able to pre-render, in any form, like static site generation, incremental site regeneration, etc, this would be a non issue...
   - One could even do this to flush a fresh site and fetch new data in the client (more easily done in pages router)
   - or the new cool technique, to flush HTML, with a suspense boundary, that resolves to your data
   - this way, nothing from your already existing server is ever public
- Careful access to DOM and other client side APIs during SSR
- Memory management, your code might cause memory leaks, or expose data through usage of globals
- Where will you deploy this, does it fit Next.js needs? will it just run as a node app? dockerized?
- Caching of certain pieces of data, done by Next.js
- Routing is bound to a file system structure (though some people hack react-router-dom in, which is just a nest of bugs waiting to happen)
- ...etc

I think, your question, the way it is presented, is a yes to both cases, a so called, it depends, so that's that. 

> Is it uncommon or unusual to use next.js in such a client-only case?

No, it is ok.

> How does it compare to using 'create react app' or 'plain react' in such a use case?

Depends, but it is ok.

**Reply by sam2000** [reply]
Thank you so much @icyJoseph 
It is very kind of you to take you time to give such a detailed reply. Your reply packs a lot of useful information and is very very helpful. 
I would like to ask some follow up questions to make sure I understand everything in full

1. Can you expand on  "Next.js is a backend for your frontend."? Does this mean Next.js can be used as a backend for some other front end? I thought it is just an "all in one" solution. I have used Ruby on Rails in the past and that seemed to do everything - render views, do the routing, talk to the data base etc. And I thought Next.js seems to be similar in that respect?
2.  I will add some more context so that you can better understand what I am trying to find out. I have built a small AWS serverless 'backend' . Basically I have a bunch of public REST endpoints that are protected by authentication. My plan is to build all the logic, database access etc on the AWS side with lambda's etc. and just expose these public endpoints. My reasoning was, if I did it this way, my API can serve all sorts of clients (mobile apps, web clients ) without being too specific to any platform. Then I have built a basic mobile client which works fine. Now I need a web client. After some research I thought react sounds cool. Then, on the react website it said something like "React is cool but if you are building an app think of using a framework like Next.js" then I started looking at Next js. But to be honest, I come from the mobile app world and I am a total new comer to modern web development.  So I wanted to double check if I am on the right track and if next js is a suitable choice for this kind of requirement.
3.  Now after seeing your reply above, I realise things are much more complicated than I thought. One of the issues is, as you pointed out, if I use next.js which is also a 'back end' then I have a kind of 'two stage backend'  The client side calls the next.js backend which calls my aws public endpoints which seem wasteful. Not sure how common it is for people to use next js in this way? as a backend that accesses another backend?
4.  You have mentioned the above issue (two stage backends) and the potential latency. You have then went on to give some possible solutions 
 

> However, if you are able to pre-render, in any form, like static site generation, incremental site regeneration, etc, this would be a non issue..
 
I guess this means we let  next js 'pre render' the page? The request goes to next js. next js then picks up a 'pre rendered' html and serves it.  I guess this only works for static content? But is this suitable for an interactive app (for example an app that serves personalised content to each user? like an app that lets users attempt practice tests that are dynamically generated and gives them immediate feedback as they attempt each question?  You still need to access the 'real backend' to fetch the right question for that user I guess?

> One could even do this to flush a fresh site and fetch new data in the client (more easily done in pages router)

Here can you explain what you meant by "flush a fresh site"?

**Reply by icyJoseph** [reply]
> 1. Can you expand on  "Next.js is a backend for your frontend."? Does this mean Next.js can be used as a backend for some other front end? I thought it is just an "all in one" solution. I have used Ruby on Rails in the past and that seemed to do everything - render views, do the routing, talk to the data base etc. And I thought Next.js seems to be similar in that respect?\

The idea is that you create a slight separation of concerns, by having a backend dedicated to your client application. Rather than an all powerful backend that serves several clients. In that view, Next.js' server, primary concern, is rendering HTML, allowing you to also have entry points for your client, and potentially hiding away implementation details of how data is fetched in the server side.  

> 2. I will add some more context so that you can better understand what I am trying to find out. I have built a small AWS serverless 'backend' . Basically I have a bunch of public REST endpoints that are protected by authentication. My plan is to build all the logic, database access etc on the AWS side with lambda's etc. and just expose these public endpoints. My reasoning was, if I did it this way, my API can serve all sorts of clients (mobile apps, web clients ) without being too specific to any platform. Then I have built a basic mobile client which works fine. Now I need a web client. After some research I thought react sounds cool. Then, on the react website it said something like "React is cool but if you are building an app think of using a framework like Next.js" then I started looking at Next js. But to be honest, I come from the mobile app world and I am a total new comer to modern web development.  So I wanted to double check if I am on the right track and if next js is a suitable choice for this kind of requirement.

You can still, do a whole bunch of work with Next.js and then have routes that fetch from your API on the client. Next.js just gives you more tools, that are often needed for a production application. Actually, the idea is that you wouldn't have to worry too much about these things, and just get started coding React.

Many years ago, I made this presentation to a MeetUp group, https://windy-fold.surge.sh/#/normal/0/-1, maybe you wanna go through it as well.  

> 3. Now after seeing your reply above, I realise things are much more complicated than I thought. One of the issues is, as you pointed out, if I use next.js which is also a 'back end' then I have a kind of 'two stage backend'  The client side calls the next.js backend which calls my aws public endpoints which seem wasteful. Not sure how common it is for people to use next js in this way? as a backend that accesses another backend?

It is something to evaluate in a per case basis, generally, it should be ok. One more request is often not adding that much into the total request time, but I've seen it all at this point. Imagine if your Next.js backend had to resolve graphql queries, which in turn request data from other services, which in turn have to go and fetch even more data...

> I guess this means we let next js 'pre render' the page? The request goes to next js. next js then picks up a 'pre rendered' html and serves it. I guess this only works for static content? But is this suitable for an interactive app (for example an app that serves personalised content to each user? like an app that lets users attempt practice tests that are dynamically generated and gives them immediate feedback as they attempt each question? You still need to access the 'real backend' to fetch the right question for that user I guess?

It can still be, you can pre-render certain parts, that don't change for all users, but if the data is highly dynamic, then you can still opt out from this for those routes.

> Here can you explain what you meant by "flush a fresh site"?

You pre-render a site, using the result of data-fetching functions as initial props to your state, and then through a use-effect or whatever fetching library, update state client side.

With App Router, this is also possible, and there's something called PPR coming up, which allows you pre-render everything up to a Suspense boundary, and the contents of that boundary are streamed during runtime, but the everything else around it, is already pre-rendered, so it can be sent to the client immediately.

**Reply by sumuhans** [reply]
Thanks so much for the detailed response. And the presentation was very useful too. I will continue to experiment and see who this goes. Thanks so much for your help. Very kind of you! 😊
